src/rust/engine/sharded_lmdb/src/lib.rs

```diff
@@ -378,6 +381,109 @@ impl ShardedLmdb {
       .await
   }
 
+  ///
+  /// Stores the given Read instance under its computed digest. This method performs two passes
+  /// over the source to 1) hash it, 2) store it. If !data_is_immutable, the second pass will
+  /// re-hash the data to confirm that it hasn't changed.
+  ///
+  pub async fn store<F, R>(
+    &self,
+    initial_lease: bool,
+    data_is_immutable: bool,
+    data_provider: F,
+  ) -> Result<Digest, String>
+  where
+    R: Read + Debug,
```

Comment:
Yea, good point. Requiring `Seek` would make testing slightly more challenging, but it could imagine it potentially being less costly from a syscall perspective?

Reply:
Yeah, I imagine it's slightly cheaper, but probably not a big deal... Probably not worth doing if it makes the implementation more complex, but I think it'd be similar complexity to what's currently there... I suspect the Seek impl for testing would be nice and simple though - a little helper struct that wraps a `Vec<Bytes>` and a `usize` pointing at which one is next to be read, where `Seek` just increments the `usize`

Reply:
So, I tried this out, and it didn't make it noticeable difference in performance. Adding `Seek` didn't avoid the need for the provider function, because callers couldn't block on `File::open` in async contexts. And given that it's challenging to implement `Seek` for `Bytes`/`Buf` (since they don't retain their original size, and instead walk their pointer through an anonymous underlying reference), I'm going to leave it as `Read` with the provider function.